planetiler-core/src/test/java/com/onthegomap/planetiler/util/LoopLineMergerTest.java

```diff
@@ -0,0 +1,398 @@
+    var merger = new LoopLineMerger()
+      .setMinLength(-1)
+      .setStubMinLength(-1)
+      .setTolerance(-1)
+      .setLoopMinLength(100);
+
+    merger.add(newLineString(
+      10, 10,
+      20, 10,
+      30, 10,
+      30, 20,
+      40, 20
+    ));
+    merger.add(newLineString(
+      20, 10,
+      30, 20
+    ));
+    assertEquals(
+      List.of(
+        newLineString(
+          10, 10,
+          20, 10,
+          30, 20,
+          40, 20
+        )
+      ),
+      merger.getMergedLineStrings()
+    );
+  }
+
+  @Test
+  void testRemoveSelfClosingLoops() {
+    // Note that self-closing loops are considered stubs.
+    // They are removed by stubMinLength, not loopMinLength...
+    var merger = new LoopLineMerger()
+      .setMinLength(-1)
+      .setTolerance(-1)
+      .setStubMinLength(5)
+      .setLoopMinLength(-1);
+
+    merger.add(newLineString(
+      1, -10, 
+      1, 1, 
+      1, 2, 
+      0, 2, 
+      0, 1, 
+      1, 1, 
+      10, 1));
+    assertEquals(
+      List.of(newLineString(1, -10, 1, 1, 10, 1)),
+      merger.getMergedLineStrings()
+    );
+  }
+
+  @Test
+  void testDoNotRemoveLargeLoops() {
+    var merger = new LoopLineMerger()
+      .setMinLength(-1)
+      .setLoopMinLength(0.001);
+
+    merger.add(newLineString(
+      10, 10,
+      20, 10,
+      30, 10,
+      30, 20,
+      40, 20
+    ));
+    merger.add(newLineString(
+      20, 10,
+      30, 20
+    ));
+    assertEquals(
+      List.of(
+        newLineString(
+          10, 10,
+          20, 10
+        ),
+        newLineString(
+          20, 10,
+          30, 10,
+          30, 20
+        ),
+        newLineString(
+          20, 10,
+          30, 20
+        ),
+        newLineString(
+          30, 20,
+          40, 20
+        )
+      ),
+      merger.getMergedLineStrings()
+    );
+  }
+
+  @Test
+  void testRemoveShortLine() {
+    var merger = new LoopLineMerger()
+      .setMinLength(10)
+      .setStubMinLength(-1)
+      .setTolerance(-1)
+      .setLoopMinLength(-1);
+
+    merger.add(newLineString(10, 10, 11, 11));
+    merger.add(newLineString(20, 20, 30, 30));
+    merger.add(newLineString(30, 30, 40, 40));
+    assertEquals(
+      List.of(newLineString(20, 20, 30, 30, 40, 40)),
+      merger.getMergedLineStrings()
+    );
+  }
+
+  @Test
+  void testRemovesShortStubsTheNonStubsThatAreTooShort() {
+    var merger = new LoopLineMerger()
+      .setMinLength(0)
+      .setLoopMinLength(-1)
+      .setStubMinLength(15)
+      .setTolerance(-1);
+
+    merger.add(newLineString(0, 0, 20, 0));
+    merger.add(newLineString(20, 0, 30, 0));
+    merger.add(newLineString(30, 0, 50, 0));
+    merger.add(newLineString(20, 0, 20, 10));
+    merger.add(newLineString(30, 0, 30, 10));
+
+    assertEquals(
+      List.of(newLineString(0, 0, 20, 0, 30, 0, 50, 0)),
+      merger.getMergedLineStrings()
+    );
+  }
+
+  @Test
+  void testMergeCarriagewaysWithOneSplitShorterThanLoopMinLength() {
```

Comment:
for the merge carriageways tests I added, I think we just need mergeStrokes=true, and tolerance=-1 (by changing the default value) for them to pass.  We may also need to change the output line order.